db/migrations/20230501205418_cache.sql

```diff
@@ -0,0 +1,39 @@
+-- migrate:up
+
+CREATE TABLE IF NOT EXISTS swoop.input_items (
+  item_uuid uuid NOT NULL DEFAULT gen_random_uuid(),
+  item_id text,
+  collection text
+);
+
+CREATE INDEX ON swoop.input_items (item_uuid);
+
+
+CREATE TABLE IF NOT EXISTS swoop.item_payload (
+  item_uuid uuid NOT NULL,
+  payload_uuid uuid NOT NULL
```

Comment:
Maybe we're saying the same thing here in two different ways, I can't tell. So let me offer an example. If we have:

```
# input_item
item_id: 'an-id'
collection: 'collection'

# item_payload
item_id: 'an-id'
collection: 'collection'
payload_uuid: '98f94fa2-2265-4038-8d8a-ca6cffe35d93'

# payload_cache
payload_uuid: '98f94fa2-2265-4038-8d8a-ca6cffe35d93'
...
```

Then with

```
    FOREIGN KEY (
      item_id,
      collection
    ) REFERENCES swoop.input_items ON DELETE RESTRICT,
```

if we try to delete the input item record, that delete will be stopped early in the transaction by the `RESTRICT`.

Nothing actually prevents a user from deleting item payload rows with what I propose, because it holds the foreign keys to the other records. As far as I know, to prevent that one would need to add a trigger function on delete for each row, to check if the row's payload record still exists or not (see [this stackoverflow for an example](https://dba.stackexchange.com/questions/278440/is-there-a-way-to-allow-only-deletes-from-cascades-only)).

One plus of a delete trigger like this would be that could extend it with a query to clean up items that are no longer referenced by any item payload rows. But a trigger per row on delete makes deletes expensive, and we can write a much more performant bulk delete query for orphaned items if we ever need to clean up that table. Thus I am not sure if the benefits outweigh the cost.